src/test/test-copy.c

```diff
@@ -436,6 +436,72 @@ TEST_RET(copy_holes) {
         return 0;
 }
 
+TEST_RET(copy_holes_with_gaps) {
+        _cleanup_(unlink_tempfilep) char fn[] = "/var/tmp/test-copy-hole-with-gaps-fd-src-XXXXXX";
+        _cleanup_(unlink_tempfilep) char fn_copy[] = "/var/tmp/test-copy-hole-with-gaps-fd-tgt-XXXXXX";
+        int r, fd, fd_copy;
+        uint64_t i;
+        struct stat stat;
+        off_t blksz;
+        char data[16] = "0123456789ABCDEF";
+
+        fd = mkostemp_safe(fn);
+        assert_se(fd >= 0);
+
+        fd_copy = mkostemp_safe(fn_copy);
+        assert_se(fd_copy >= 0);
+
+        assert_se(fstat(fd, &stat) >= 0);
+        blksz = stat.st_blksize;
+
+        /* Create a file with:
+         *  - hole of 1 block
+         *  - data of 1 block
+         *  - hole of 2 blocks
+         *  - data of 1 block
+         *
+         * Since sparse files are based on blocks and not bytes, we need to make
+         * sure that the holes are aligned to the block size.
+         */
+
+        r = RET_NERRNO(fallocate(fd, FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE, 0, blksz));
+        if (ERRNO_IS_NOT_SUPPORTED(r))
+                return log_tests_skipped("Filesystem doesn't support hole punching");
+        assert_se(r >= 0);
+        assert_se(lseek(fd, blksz, SEEK_CUR) >= 0);
+        for (i = 0; i < blksz / sizeof(data); i++)
+                assert_se(write(fd, data, sizeof(data)) >= 0);
```

Comment:
We probably want to assert that we wrote as much data as we passed in here? Might be better to use `loop_write()` here and then `assert >= 0` will be sufficient